clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java

```diff
@@ -2840,6 +2840,221 @@ public void testListOffsets() throws Exception {
         }
     }
 
+    /**
```

Comment:
All the tests here deal with single node clusters, whereas in real production Kafka environments, there are many more. I think in order to catch the bug mentioned in my previous comment, it would be good to include a test that tests a multi node Kafka cluster and to list offsets from that, where each topic partition has a different leader.

Reply:
Yes. I added a test case with 3 TP with 3 leaders. The test case mock error response for only one of the TP. Ideally, we need to mock error responses for all the 3 TP but this would cause race condition in the test framework. Since the test framework is pushing mock response in the queue and the IO is implemented by "react pattern", we probably can't determine the actual order for the Metadata Call and ListOffset Call before run time. But the test case can verify that the list offset call is per leader. Let me know if you can think of a better way of testing.

Reply:
testListOffsetsMaxAllowedNumRetriesSuccessWithMultipleLeaders

Reply:
Wouldn't the metadata request always happen before the failed calls are executed again? I believe the logic flow goes:
1) Metadata Request
2) ListOffsets --> 2/3 failed partitions
3) Metadeta Request
3) ListOffsets --> success

Instead of mocking all 3 TP, how about just mocking 2 out of the 3? You can use some inspiration from the test here: https://github.com/apache/kafka/blob/trunk/clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java#L2997-L3057

This test mocks a 3 node cluster with 2 topic partitions which initially start out to have the same partition leader and then mocks a retriable error and both partitions change to have a different leader.

What I was more thinking for your patch was to recreate the same exact testing logic you have for the single node clusters for the multi-node clusters. So in a multi-node clusters with varying topic partitions, testing that the patch you have now respects the retry config and the backoff config for each call made. An example test could be you have 3 different partitions that retry different amount of times (one partition retries once, another retries twice, and yet another retries thrice), and let's say the max retries is 2, then the call associated with the partition that is trying to retry a third time should fail. Along those lines, you could also test having a multi-node cluster that tests retry-backoff in the same way.

Reply:
I just reviewed your new test `testListOffsetsMaxAllowedNumRetriesSuccessWithMultipleLeaders` and it looks good! Overall happy with the test coverage introduced here.

Reply:
Thanks.